packages/integrations/mdx/src/internal-utils.ts

```diff
@@ -0,0 +1,104 @@
+import type { Options as AcornOpts } from 'acorn';
+import { parse } from 'acorn';
+import type { AstroConfig, SSRError } from 'astro';
+import matter from 'gray-matter';
+import type { MdxjsEsm } from 'mdast-util-mdx';
+
+function appendForwardSlash(path: string) {
+	return path.endsWith('/') ? path : path + '/';
+}
+
+interface FileInfo {
+	fileId: string;
+	fileUrl: string;
+}
+
+/** @see 'vite-plugin-utils' for source */
+export function getFileInfo(id: string, config: AstroConfig): FileInfo {
+	const sitePathname = appendForwardSlash(
+		config.site ? new URL(config.base, config.site).pathname : config.base
+	);
+
+	// Try to grab the file's actual URL
+	let url: URL | undefined = undefined;
+	try {
+		url = new URL(`file://${id}`);
+	} catch {}
+
+	const fileId = id.split('?')[0];
+	let fileUrl: string;
+	const isPage = fileId.includes('/pages/');
+	if (isPage) {
+		fileUrl = fileId.replace(/^.*?\/pages\//, sitePathname).replace(/(\/index)?\.mdx$/, '');
+	} else if (url && url.pathname.startsWith(config.root.pathname)) {
+		fileUrl = url.pathname.slice(config.root.pathname.length);
+	} else {
+		fileUrl = fileId;
+	}
+
+	if (fileUrl && config.trailingSlash === 'always') {
+		fileUrl = appendForwardSlash(fileUrl);
+	}
+	return { fileId, fileUrl };
+}
+
+/**
+ * Match YAML exception handling from Astro core errors
+ * @see 'astro/src/core/errors.ts'
+ */
+export function parseFrontmatter(code: string, id: string) {
+	try {
+		return matter(code);
+	} catch (e: any) {
+		if (e.name === 'YAMLException') {
+			const err: SSRError = e;
+			err.id = id;
+			err.loc = { file: e.id, line: e.mark.line + 1, column: e.mark.column };
+			err.message = e.reason;
+			throw err;
+		} else {
+			throw e;
+		}
+	}
+}
+
+/**
+ * Convert a string of JavaScript to an MDX-compatible tree node
+ * for use in remark or rehype plugins
+ * @param jsString String of valid JavaScript
+ * @param acornOpts Acorn parser options ([see Acorn docs](https://github.com/acornjs/acorn/tree/master/acorn#interface) for more).
+ * @returns An MDX-compatible tree node
+ */
+export function jsToMdxTreeNode(
+	jsString: string,
+	acornOpts: AcornOpts = {
+		ecmaVersion: 'latest',
+		sourceType: 'module',
+	}
+): MdxjsEsm {
+	return {
+		type: 'mdxjsEsm',
+		value: '',
```

Comment:
This should be your jsString. Not strictly required, but improves debuggability.